File: cloud_testing/cloud_api_poc/src/main/java/cloud_api_poc/TestRunner.java
package cloud_api_poc;

import static cloud_api_poc.Utils.sleep;
import static java.lang.System.currentTimeMillis;

import com.google.testing.Testing.Projects.TestMatrices.Create;
import com.google.testing.model.AndroidMatrix;
import com.google.testing.model.TestExecution;
import com.google.testing.model.TestMatrix;
import com.google.testing.model.ToolResultsStep;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;

public abstract class TestRunner {
  private TestRunner() {}

  private static final long POLLING_TIMEOUT = 60 * 60 * 1000; // 60m
  private static final int POLLING_INTERVAL = 10 * 1000; // 10s
  private static final String FINISHED = "FINISHED";
  private static final String SUCCESS = "success";

  @SuppressWarnings("unchecked")
  public static boolean pollTests(ArrayList<String> testMatrixIds) {
    long stopTime = currentTimeMillis() + POLLING_TIMEOUT;
    List<String> waitingOnIds = (List<String>) testMatrixIds.clone();
    int matrixCount = testMatrixIds.size();
    List<TestMatrix> finishedTestMatrix = new ArrayList<>(matrixCount);

    boolean invalid = false;
    String lastStatus = "";
    String currentStatus;
    String lastState = "";
    TestExecution lastTestExecution = null;

    while (waitingOnIds.size() > 0) {
      if (currentTimeMillis() > stopTime) {
        throw new RuntimeException(
            "Polling timed out after "
                + TimeUnit.MILLISECONDS.toMinutes(POLLING_TIMEOUT)
                + " minutes");
      }

      currentStatus = finishedTestMatrix.size() + " / " + matrixCount + " complete";
      if (!currentStatus.equals(lastStatus)) {
        System.out.println(currentStatus);
      }
      lastStatus = currentStatus;

      String testMatrixId = waitingOnIds.get(0);
      TestMatrix testMatrix = GcTestMatrix.refresh(testMatrixId);

      boolean finished = true;
      for (TestExecution testExecution : testMatrix.getTestExecutions()) {
        lastState = testExecution.getState();
        System.out.println(testExecution.getState());

        if (!TestExecutionState.isValid(lastState)) {
          invalid = true;
          lastTestExecution = testExecution;
          break;
        }

        if (!FINISHED.equals(lastState)) {
          finished = false;
          break;
        }
      }

      if (invalid) {
        break;
      }

      if (finished) {
        finishedTestMatrix.add(testMatrix);
        waitingOnIds.remove(0);
      } else {
        sleep(POLLING_INTERVAL);
      }
    }

    if (lastState == null) lastState = "";

    if (invalid) {
      // infrastructure errors are transparently retried up to 3 attempts.
      if (lastState.equals(TestExecutionState.ERROR) && lastTestExecution != null) {
        System.out.println(lastTestExecution.getTestDetails().getErrorMessage());
      }
      for (String matrixId : waitingOnIds) {
        GcTestMatrix.cancel(matrixId);
      }

      // Invalid state detected: ERROR is an Infrastructure failure.
      throw new RuntimeException("Invalid state detected: " + lastState);
    }

    System.out.println("Test run finished");
    currentStatus = finishedTestMatrix.size() + " / " + matrixCount + " complete.";
    System.out.println(currentStatus);
    System.out.println();

    System.out.println("Fetching test results...");
    List<ToolResultsValue> results = fetchTestResults(finishedTestMatrix);

    boolean allTestsSuccessful = true;
    long billableMinutes = 0;
    long successCount = 0;
    long failureCount = 0;
    long totalCount = testMatrixIds.size();
    for (ToolResultsValue result : results) {
      System.out.println(result);
      System.out.println();

      if (result.outcome.equals(SUCCESS)) {
        successCount += 1;
      } else {
        allTestsSuccessful = false;
        failureCount += 1;
      }

      billableMinutes += result.billableMinutes;
    }

    System.out.println(
        totalCount + " tests. successful: " + successCount + ", failed: " + failureCount);
    Billing.estimateCosts(billableMinutes);

    return allTestsSuccessful;
  }

  @SuppressWarnings("unchecked")
  private static List<ToolResultsValue> fetchTestResults(List<TestMatrix> finishedTestMatrix) {
    List<ToolResultsStep> steps = new ArrayList<>();
    for (TestMatrix testMatrix : finishedTestMatrix) {
      for (TestExecution testExecution : testMatrix.getTestExecutions()) {
        steps.add(testExecution.getToolResultsStep());
      }
    }

    int shardCount = steps.size();
    Parallel parallel = new Parallel(shardCount);
    List<ToolResultsValue> toolValues = Collections.synchronizedList(new ArrayList(shardCount));

    for (int i = 0; i < shardCount; i++) {
      parallel.addCallable(new ToolResultsCallable(toolValues, steps.get(i)));
    }

    parallel.run();

    if (shardCount != toolValues.size() || shardCount != toolValues.size()) {
      throw new RuntimeException("Synchronization error");
    }

    return toolValues;
  }

  public static ArrayList<String> scheduleTests(
      String appApkGcsPath, String testApkGcsPath, List<String> testMethodNames) {

    int shardCount = testMethodNames.size();

    // *MUST* use synchronized list to play nice with ExecutorService
    List<String> testMatrixIds = Collections.synchronizedList(new ArrayList(shardCount));
    Parallel parallel = new Parallel(shardCount);

    for (String testMethod : testMethodNames) {
      AndroidMatrix androidMatrix = GcAndroidMatrix.build("NexusLowRes", "25", "en", "portrait");

      Create testMatrixCreate =
          GcTestMatrix.build(appApkGcsPath, testApkGcsPath, androidMatrix, testMethod);

      parallel.addCallable(new MatrixCallable(testMatrixIds, testMatrixCreate));
    }

    parallel.run();

    if (shardCount != testMatrixIds.size() || shardCount != testMatrixIds.size()) {
      throw new RuntimeException("Synchronization error");
    }

    return new ArrayList<>(testMatrixIds);
  }
}

File: cloud_testing/kotlin_poc/src/main/kotlin/ftl/TestRunner.kt
package ftl

import com.google.cloud.storage.Storage
import com.google.cloud.storage.StorageOptions
import com.google.testing.model.TestExecution
import com.google.testing.model.TestMatrix
import ftl.GcStorage.uploadApk
import ftl.Utils.sleep
import java.lang.System.currentTimeMillis
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*
import java.util.concurrent.TimeUnit

object TestRunner {

    private const val POLLING_TIMEOUT = (60 * 60 * 1000).toLong() // 60m
    private const val POLLING_INTERVAL = 10 * 1000 // 10s
    private const val FINISHED = "FINISHED"
    private const val SUCCESS = "success"
    private val storage = StorageOptions.newBuilder().build().service
    private val testResultRgx = Regex(".*test_result_\\d+.xml$")

    // gcsFolder = "2018-01-17_19:38:56.695000_fCMj"
    private fun downloadXml(gcsFolder: String) {
        if (!GlobalConfig.downloadXml) return
        var xmlCount = -1

        val prefix = Storage.BlobListOption
                .prefix(gcsFolder)

        val fields = Storage.BlobListOption.fields(Storage.BlobField.NAME)
        val result = storage.list(GlobalConfig.bucketGcsPath, prefix, fields)

        result.iterateAll().forEach({
            val name = it.blobId.name
            if (name.matches(testResultRgx)) {
                xmlCount += 1
                println("Downloading: $name")
                it.downloadTo(Paths.get(gcsFolder, "test_result_$xmlCount.xml"))
            }
        })
    }

    fun pollTests(testMatrixIds: ArrayList<String>): Boolean {
        val stopTime = currentTimeMillis() + POLLING_TIMEOUT
        val waitingOnIds = ArrayList<String>(testMatrixIds)
        val matrixCount = testMatrixIds.size
        val finishedTestMatrixArray = ArrayList<TestMatrix>(matrixCount)

        var invalid = false
        var lastStatus = ""
        var currentStatus: String
        var lastState: String? = ""
        var lastTestExecution: TestExecution? = null

        while (waitingOnIds.size > 0) {
            if (currentTimeMillis() > stopTime) {
                throw RuntimeException(
                        "Polling timed out after "
                                + TimeUnit.MILLISECONDS.toMinutes(POLLING_TIMEOUT)
                                + " minutes")
            }

            currentStatus = finishedTestMatrixArray.size.toString() + " / " + matrixCount + " complete"
            if (currentStatus != lastStatus) {
                println(currentStatus)
            }
            lastStatus = currentStatus

            val testMatrixId = waitingOnIds[0]
            val testMatrix = GcTestMatrix.refresh(testMatrixId)

            var finished = true
            for (testExecution in testMatrix.testExecutions) {
                lastState = testExecution.state
                println(testExecution.state)

                if (!TestExecutionState.isValid(lastState)) {
                    // TEST_ONLY_APK: The APK is marked as "testOnly".
                    // Use Build > Build APK(s) in Android Studio to build the app apk.
                    // App APKs built from using 'Run' in Android Studio don't work on Firebase Test Lab
                    println("Error: " + testExecution.testDetails.errorMessage)
                    invalid = true
                    lastTestExecution = testExecution
                    break
                }

                if (FINISHED != lastState) {
                    finished = false
                    break
                }
            }

            if (invalid) {
                break
            }

            if (finished) {
                finishedTestMatrixArray.add(testMatrix)
                waitingOnIds.removeAt(0)
            } else {
                sleep(POLLING_INTERVAL)
            }
        }

        if (lastState == null) lastState = ""

        if (invalid) {
            // infrastructure errors are transparently retried up to 3 attempts.
            if (lastState == TestExecutionState.ERROR && lastTestExecution != null) {
                println(lastTestExecution.testDetails.errorMessage)
            }
            for (matrixId in waitingOnIds) {
                GcTestMatrix.cancel(matrixId)
            }

            // Invalid state detected: ERROR is an Infrastructure failure.
            throw RuntimeException("Invalid state detected: " + lastState)
        }

        println("Test run finished")
        currentStatus = finishedTestMatrixArray.size.toString() + " / " + matrixCount + " complete."
        println(currentStatus)
        println()

        println("Fetching test results...")
        val results = fetchTestResults(finishedTestMatrixArray)

        var allTestsSuccessful = true
        var billableMinutes: Long = 0
        var successCount: Long = 0
        var failureCount: Long = 0
        val totalCount = testMatrixIds.size.toLong()
        for (result in results) {
            println(result)
            println()

            // What gcs object is associated with this result?
            // "executionId" -> "6296919122126826652"
            // "historyId" -> "bh.58317d9cd7ab9ba2"
            // "projectId" -> "delta-essence-114723"
            // "stepId" -> "bs.e7bbe13221894d52"
            if (result.outcome == SUCCESS) {
                successCount += 1
            } else {
                allTestsSuccessful = false
                failureCount += 1

                // Only download XML on failed runs
                downloadXml(result.gcsPath)
            }

            billableMinutes += result.billableMinutes
        }

        println(
                totalCount.toString() + " tests. successful: " + successCount + ", failed: " + failureCount)
        Billing.estimateCosts(billableMinutes)

        return allTestsSuccessful
    }

    private fun fetchTestResults(finishedTestMatrix: List<TestMatrix>): List<ToolResultsValue> {
        val steps = ArrayList<ToolResultsStepGcsPath>()
        for (testMatrix in finishedTestMatrix) {
            val gcsPath = testMatrix.resultStorage.googleCloudStorage.gcsPath.split('/')[1]

            testMatrix.testExecutions
                    .map { it.toolResultsStep }
                    .mapTo(steps) { ToolResultsStepGcsPath(it, gcsPath) }
        }

        val shardCount = steps.size
        val parallel = Parallel(shardCount)
        val toolValues = Collections.synchronizedList<ToolResultsValue>(ArrayList(shardCount))

        for (step in steps) {
            parallel.addCallable(ToolResultsCallable(toolValues, step))
        }

        parallel.run()

        if (shardCount != toolValues.size || shardCount != toolValues.size) {
            throw RuntimeException("Synchronization error")
        }

        return toolValues
    }

    fun scheduleApks(appApk: Path, testApk: Path, shardCount: Int, runConfig: RunConfig): ArrayList<String> {
        val appApkGcsPath = uploadApk(appApk)
        val testApkGcsPath = uploadApk(testApk)

        // *MUST* use synchronized list to play nice with ExecutorService
        val testMatrixIds = Collections.synchronizedList<String>(ArrayList(shardCount))
        val parallel = Parallel(shardCount)

        println("Running ${shardCount}x")
        repeat(shardCount) {
            val androidMatrix = GcAndroidMatrix.build("NexusLowRes", "26", "en", "portrait")

            val testMatrixCreate = GcTestMatrix.build(appApkGcsPath, testApkGcsPath, androidMatrix, runConfig = runConfig)

            parallel.addCallable(MatrixCallable(testMatrixIds, testMatrixCreate))
        }

        parallel.run()

        if (shardCount != testMatrixIds.size || shardCount != testMatrixIds.size) {
            throw RuntimeException("Synchronization error")
        }

        return ArrayList(testMatrixIds)
    }

    /** Runs 1 test per VM **/
    fun scheduleTests(
            appApkGcsPath: String, testApkGcsPath: String, testMethodNames: List<String>, runConfig: RunConfig): ArrayList<String> {

        val shardCount = testMethodNames.size

        // *MUST* use synchronized list to play nice with ExecutorService
        val testMatrixIds = Collections.synchronizedList<String>(ArrayList(shardCount))
        val parallel = Parallel(shardCount)

        for (testMethod in testMethodNames) {
            val androidMatrix = GcAndroidMatrix.build("NexusLowRes", "25", "en", "portrait")

            val testMatrixCreate = GcTestMatrix.build(appApkGcsPath, testApkGcsPath, androidMatrix, testMethod, runConfig = runConfig)

            parallel.addCallable(MatrixCallable(testMatrixIds, testMatrixCreate))
        }

        parallel.run()

        if (shardCount != testMatrixIds.size || shardCount != testMatrixIds.size) {
            throw RuntimeException("Synchronization error")
        }

        return ArrayList(testMatrixIds)
    }
}
